keras_contrib/layers/crf.py

```diff
@@ -0,0 +1,542 @@
+                  'kernel_constraint': constraints.serialize(self.kernel_constraint),
+                  'chain_constraint': constraints.serialize(self.chain_constraint),
+                  'boundary_constraint': constraints.serialize(self.boundary_constraint),
+                  'bias_constraint': constraints.serialize(self.bias_constraint),
+                  'input_dim': self.input_dim,
+                  'unroll': self.unroll}
+        base_config = super(CRF, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+    @property
+    def loss_function(self):
+        if self.learn_mode == 'join':
+            def loss(y_true, y_pred):
+                assert self.inbound_nodes, 'CRF has not connected to any layer.'
+                assert not self.outbound_nodes, 'When learn_model="join", CRF must be the last layer.'
+                if self.sparse_target:
+                    y_true = K.one_hot(K.cast(y_true[:, :, 0], 'int32'), self.units)
+                X = self.inbound_nodes[0].input_tensors[0]
+                mask = self.inbound_nodes[0].input_masks[0]
+                nloglik = self.get_negative_log_likelihood(y_true, X, mask)
+                return nloglik
+            return loss
+        else:
+            if self.sparse_target:
+                return sparse_categorical_crossentropy
+            else:
+                return categorical_crossentropy
+
+    @property
+    def accuracy(self):
+        if self.test_mode == 'viterbi':
+            return self.viterbi_acc
+        else:
+            return self.marginal_acc
+
+    @staticmethod
+    def _get_accuracy(y_true, y_pred, mask, sparse_target=False):
+        y_pred = K.argmax(y_pred, -1)
+        if sparse_target:
+            y_true = K.cast(y_true[:, :, 0], K.dtype(y_pred))
+        else:
+            y_true = K.argmax(y_true, -1)
+        judge = K.cast(K.equal(y_pred, y_true), K.floatx())
+        if mask is None:
+            return K.mean(judge)
+        else:
+            mask = K.cast(mask, K.floatx())
+            return K.sum(judge * mask) / K.sum(mask)
+
+    @property
+    def viterbi_acc(self):
+        def acc(y_true, y_pred):
+            X = self.inbound_nodes[0].input_tensors[0]
+            mask = self.inbound_nodes[0].input_masks[0]
+            y_pred = self.viterbi_decoding(X, mask)
+            return self._get_accuracy(y_true, y_pred, mask, self.sparse_target)
+        acc.func_name = 'viterbi_acc'
+        return acc
+
+    @property
+    def marginal_acc(self):
+        def acc(y_true, y_pred):
+            X = self.inbound_nodes[0].input_tensors[0]
+            mask = self.inbound_nodes[0].input_masks[0]
+            y_pred = self.get_marginal_prob(X, mask)
+            return self._get_accuracy(y_true, y_pred, mask, self.sparse_target)
+        acc.func_name = 'marginal_acc'
+        return acc
+
+    @staticmethod
+    def log_sum_exp(x, axis=-1):
```

Comment:
Then I probably should just change it.  Actually, there is a bug when using `K.logsumexp` with` `tensorflow` backend, when passed a placeholder, it losses the shape information, i.e., `K.int_shape` got None. This is due to the fact that `K.logsumexp` uses an deprecidated argument. I'll make a PR to `keras` and will make the change here one it is merged.